src/lib/autoinstall/widgets/ask_dialog.rb

```diff
@@ -0,0 +1,434 @@
+# more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, contact SUSE LLC.
+#
+# To contact SUSE LLC about this file by physical or electronic mail, you may
+# find current contact information at www.suse.com.
+
+require "yast"
+require "cwm"
+require "cwm/popup"
+
+Yast.import "Popup"
+
+module Y2Autoinstall
+  module Widgets
+    # Implements a CWM dialog for the {Y2Autoinstall::Ask::Dialog} class
+    #
+    # The dialog displays all the questions included in the dialog and updates
+    # their values accordingly.
+    #
+    # @example 'Running' a dialog
+    #   question = Y2Autoinstall::Ask::Question.new("Username")
+    #   dialog = Y2Autoinstall::Ask::Dialog.new(1, [question])
+    #   AskDialog.new(dialog).run
+    class AskDialog < CWM::Popup
+      # Common methods for question widgets
+      module AskWidget
+        # Returns the widget_id, which is based in the type and the question's object_id
+        #
+        # @return [String] Widget ID
+        def widget_id
+          "#{widget_type}_#{@question.object_id}"
+        end
+
+        # Returns the widget's label
+        #
+        # The widget's label corresponds to the question text.
+        #
+        # @return [String]
+        # @macro seeAbstractWidget
+        def label
+          @question.text
+        end
+
+        # Initializes the widget value
+        #
+        # Use the question's value or the default one.
+        # @macro seeAbstractWidget
+        def init
+          if !@question.value.nil?
+            self.value = @question.value.to_s
+            return
+          end
+
+          if @question.default_value_script
+            default_from_script = run_script(@question.default_value_script)
+          end
+          self.value = default_from_script || @question.default.to_s
+        end
+
+        # Stores the widget's value in the question
+        #
+        # @macro seeAbstractWidget
+        def store
+          @question.value = value
+        end
+
+      private
+
+        def run_script(script)
+          script.create_script_file
+          script.execute
+        end
+      end
+
+      # CheckBox widget for a question
+      class CheckBox < CWM::CheckBox
+        include AskWidget
+
+        # @param question [Y2Autoinstall::Ask::Question] Question to represent
+        def initialize(question)
+          textdomain "autoinst"
+          @question = question
+        end
+
+        # @macro seeAbstractWidget
+        def opt
+          [:notify]
+        end
+
+        # @macro seeAbstractWidget
+        def init
+          if !@question.value.nil?
+            self.value = @question.value
+            return
+          end
+
+          if @question.default_value_script
+            default_from_script = run_script(@question.default_value_script)
+          end
+          self.value = default_from_script == "true" || @question.default == "true"
```

Comment:
I know that it is a bit corner case but here can happen that script return false and default can be "true". In documentation I see nothing that explain what has precedence or that it is not supported to specify both.

Reply:
Actually, I have checked the behavior again and the script has precedence unless it fails. In that case, it falls back to the value specified by the `<default>` element.